barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -66,50 +66,94 @@ void AvmTraceBuilder::reset()
 
     external_call_counter = 0;
 }
-
-AvmTraceBuilder::IndirectThreeResolution AvmTraceBuilder::resolve_ind_three(
-    uint8_t space_id, uint32_t clk, uint8_t indirect, uint32_t a_offset, uint32_t b_offset, uint32_t c_offset)
+/**
+ * @brief Returns an array of mem_offsets and tags them with their given Addressing Mode (direct/indirect) based on the
+ * given indirect byte.
+ * @tparam N The number of memory offsets to resolve.
+ */
+template <size_t N> std::array<ResolvedAddr, N> resolve_indirects(uint8_t indirect, std::array<uint32_t, N> mem_offsets)
 {
-    bool indirect_flag_a = is_operand_indirect(indirect, 0);
-    bool indirect_flag_b = is_operand_indirect(indirect, 1);
-    bool indirect_flag_c = is_operand_indirect(indirect, 2);
-
-    uint32_t direct_a_offset = a_offset;
-    uint32_t direct_b_offset = b_offset;
-    uint32_t direct_c_offset = c_offset;
-
-    bool tag_match = true;
-
-    if (indirect_flag_a) {
-        auto read_ind_a =
-            mem_trace_builder.indirect_read_and_load_from_memory(space_id, clk, IndirectRegister::IND_A, a_offset);
-        direct_a_offset = uint32_t(read_ind_a.val);
-        tag_match = tag_match && read_ind_a.tag_match;
+    std::array<ResolvedAddr, N> resolved_addresses;
+
+    for (size_t i = 0; i < N; i++) {
+        // No need to type this as a bool as is implied by the (& 1).
+        uint8_t indirect_bit = (indirect >> i) & 1;
+        // Cast straight to AddressingMode, saves having to have a branching statement here.
+        auto addr = static_cast<AddressingMode>(indirect_bit);
+        resolved_addresses[i] = std::make_pair(addr, mem_offsets[i]);
     }
+    return resolved_addresses;
+}
 
-    if (indirect_flag_b) {
-        auto read_ind_b =
-            mem_trace_builder.indirect_read_and_load_from_memory(space_id, clk, IndirectRegister::IND_B, b_offset);
-        direct_b_offset = uint32_t(read_ind_b.val);
-        tag_match = tag_match && read_ind_b.tag_match;
+/**
+ * @brief Loads a value from memory into a given intermediate register at a specified clock cycle.
+ * Handles both direct and indirect memory access.
+ * @tparam reg The intermediate register to load the value into.
+ */
+template <IntermRegister reg>
+AvmTraceBuilder::MemOp AvmTraceBuilder::constrained_read_from_memory(
+    uint8_t space_id, uint32_t clk, ResolvedAddr addr, AvmMemoryTag read_tag, AvmMemoryTag write_tag)
+{
+    // Get the same matching indirect register for the given intermediate register.
```

Comment:
Not for this PR, but this is just converting ints; not very strong type-wise

Reply:
I strongly suggest a little mapping converter from Intermregister to IndirectRegister. This way too hacky in my opinion.

Reply:
Totally agree. The mapping of registers is also what I had in mind in the future